File: StockSpanProblem/StockSpanProblem.cpp
/*
https://www.geeksforgeeks.org/problems/stock-span-problem-1587115621/1
*/

class Solution {
  public:
    vector<int> calculateSpan(vector<int>& arr) {
        int sz=arr.size();
        vector<int>res;
        stack<int>stk;
        for(int i=0; i<sz; ++i){
            int currprice=arr[i];
            while(!stk.empty()&&arr[stk.top()]<=currprice) stk.pop();
            int curcount=stk.empty()?i+1:(i-stk.top());
            res.push_back(curcount);
            stk.push(i);
        }
        return res;
    }
};

File: StockSpanProblem/StockSpanProblem.java
/*
https://www.geeksforgeeks.org/problems/stock-span-problem-1587115621/1
*/

class Solution {
    public ArrayList<Integer> calculateSpan(int[] arr) {
        int lengthOfArr=arr.length;
        ArrayList<Integer>resultantSpan=new ArrayList<>();
        Stack<Integer>stackOfIndices=new Stack<>();
        for(int i=0; i<lengthOfArr; ++i){
            int currentPrice=arr[i];
            while(!stackOfIndices.isEmpty()&&arr[stackOfIndices.peek()]<=currentPrice){
                stackOfIndices.pop();
            }
            int currentCount=stackOfIndices.isEmpty()?i+1:(i-stackOfIndices.peek());
            resultantSpan.add(currentCount);
            stackOfIndices.add(i);
        }
        return resultantSpan;
    }
}